### PR TITLE
4.x Adding strict types to the Http Client

### PR DESCRIPTION
**Pull request by burzum**
#11935

Before merging this please take a look at the _getBody() and bod() methods. They're inconsistent and because of this I haven't yet added a return type there. I'm not sure if the code should be corrected or the doc blocks.

### CODE REVIEW

File: src/Http/Client/Adapter/Stream.php [hunk 1]
```diff
@@ -114,14 +115,14 @@ public function createResponses($headers, $content)
...
-    protected function _buildContext(Request $request, $options)
+    protected function _buildContext(Request $request, array $options): void
     {
         $this->_buildContent($request, $options);
         $this->_buildHeaders($request, $options);
         $this->_buildOptions($request, $options);
 
         $url = $request->getUri();
-        $scheme = parse_url($url, PHP_URL_SCHEME);
+        $scheme = parse_url((string)$url, PHP_URL_SCHEME);
```
Comment (ADmad) [inline src/Http/Client/Adapter/Stream.php]: This makes me sad. When a class has `__toString()` method explicit type casting shouldn't be necessary.

File: src/Http/Client/FormData.php [hunk 2]
```diff
@@ -239,7 +240,7 @@ public function contentType()
      *
      * @return string
      */
-    public function __toString()
+    public function __toString(): string
```
Comment (ADmad) [inline src/Http/Client/FormData.php]: It would be best to avoid adding type hint's to PHP's magic methods to be safe for future.

**Comment by ADmad** [comment]
I pushed some updates.

File: src/Http/Client/FormData.php (hunk 2 above)
Comment (dereuromark) [reply src/Http/Client/FormData.php]: What do you mean?

File: src/Http/Client/Adapter/Stream.php (hunk 1 above)
Comment (burzum) [reply src/Http/Client/Adapter/Stream.php]: How do you know it has __toString() What if the implementation of the object changes? We can ensure this method exists on the object.

File: src/Http/Client/FormData.php (hunk 2 above)
Comment (markstory) [reply src/Http/Client/FormData.php]: I wouldn't expect PHP to change the contract on `__toString` any time soon.

**Review by markstory — APPROVED** [review]
Changes look good to me.

**Pull request merged**